From 022337ecdb430a2e536db8dbd3199b9d3e3cf58e Mon Sep 17 00:00:00 2001
From: Zbigniew Lukwinski <zbigniew.lukwinski@intel.com>
Date: Thu, 13 Mar 2025 15:18:16 +0100
Subject: [PATCH] Support x86_64_unknown target

Signed-off-by: Zbigniew Lukwinski <zbigniew.lukwinski@intel.com>
---
 Cargo.toml |  2 +-
 build.rs   | 18 +++++++++++++++++-
 2 files changed, 18 insertions(+), 2 deletions(-)

diff --git a/Cargo.toml b/Cargo.toml
index 8f8da17da..a45ec829d 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -159,7 +159,7 @@ name = "ring"
 
 [dependencies]
 cfg-if = { version = "1.0.0", default-features = false }
-getrandom = { version = "0.2.10" }
+getrandom = { version = "0.2.10", features = ["rdrand"] }
 untrusted = { version = "0.9" }
 
 [target.'cfg(all(any(all(target_arch = "aarch64", target_endian = "little"), all(target_arch = "arm", target_endian = "little")), any(target_os = "android", target_os = "linux")))'.dependencies]
diff --git a/build.rs b/build.rs
index 9843ad8aa..494d8cfc7 100644
--- a/build.rs
+++ b/build.rs
@@ -254,6 +254,8 @@ const LINUX_ABI: &[&str] = &[
     "linux",
     "redox",
     "solaris",
+    // For `x86_64-unknown-none` target
+    "none",
 ];
 
 const WIN32N: &str = "win32n";
@@ -594,14 +596,28 @@ fn configure_cc(c: &mut cc::Build, target: &Target, c_root_dir: &Path, include_d
     // Allow cross-compiling without a target sysroot for these targets.
     if (target.arch == WASM32)
         || (target.os == "linux" && target.env == "musl" && target.arch != X86_64)
+        || (target.os == "none")
     {
         // TODO: Expand this to non-clang compilers in 0.17.0 if practical.
         if compiler.is_like_clang() {
             let _ = c.flag("-nostdlibinc");
+            // Required on windows for cross compilation to `x86_64-unknown-none`
+            let _ = c.flag("-ffreestanding");
             let _ = c.define("RING_CORE_NOSTDLIBINC", "1");
         }
     }
 
+    // `clang` does not define `__ELF__` for `x86_64-unknown-none` target.
+    // Manually define it.
+    if target.os == "none" {
+        if let Ok(compiler) = c.try_get_compiler() {
+            if compiler.is_like_clang() {
+                let _ = c.define("__ELF__", None);
+                let _ = c.flag("-fPIC");
+            }
+        }
+    }
+
     if target.force_warnings_into_errors {
         c.warnings_into_errors(true);
     }
@@ -619,7 +635,7 @@ fn nasm(file: &Path, arch: &str, include_dir: &Path, out_dir: &Path, c_root_dir:
     let mut include_dir = include_dir.as_os_str().to_os_string();
     include_dir.push(OsString::from(String::from(std::path::MAIN_SEPARATOR)));
 
-    let mut c = Command::new("./target/tools/windows/nasm/nasm");
+    let mut c = Command::new("nasm");
     let _ = c
         .arg("-o")
         .arg(out_file.to_str().expect("Invalid path"))
diff --git a/src/arithmetic/bigint/modulus.rs b/src/arithmetic/bigint/modulus.rs
index b0cd6d6..7c7c6f4 100644
--- a/src/arithmetic/bigint/modulus.rs
+++ b/src/arithmetic/bigint/modulus.rs
@@ -113,7 +113,7 @@ impl<M> OwnedModulus<M> {
         })
     }
 
-    pub(crate) fn modulus(&self, cpu_features: cpu::Features) -> Modulus<M> {
+    pub(crate) fn modulus(&self, cpu_features: cpu::Features) -> Modulus<'_, M> {
         Modulus {
             limbs: self.inner.limbs(),
             n0: self.n0,
diff --git a/src/arithmetic/limbs512/storage.rs b/src/arithmetic/limbs512/storage.rs
index 210192b..99f2277 100644
--- a/src/arithmetic/limbs512/storage.rs
+++ b/src/arithmetic/limbs512/storage.rs
@@ -45,7 +45,7 @@ impl<const N: usize> AlignedStorage<N> {
         &mut self,
         num_entries: usize,
         chunks_per_entry: usize,
-    ) -> Result<AsChunksMut<Limb, LIMBS_PER_CHUNK>, LenMismatchError> {
+    ) -> Result<AsChunksMut<'_, Limb, LIMBS_PER_CHUNK>, LenMismatchError> {
         let total_limbs = num_entries * chunks_per_entry * LIMBS_PER_CHUNK;
         let len = self.0.len();
         let flattened = self
diff --git a/src/pkcs8.rs b/src/pkcs8.rs
index 5ef66fd..39d81f5 100644
--- a/src/pkcs8.rs
+++ b/src/pkcs8.rs
@@ -53,7 +53,7 @@ pub(crate) struct Template {
 
 impl Template {
     #[inline]
-    fn alg_id_value(&self) -> untrusted::Input {
+    fn alg_id_value(&self) -> untrusted::Input<'_> {
         untrusted::Input::from(self.alg_id_value_())
     }
 
@@ -62,7 +62,7 @@ impl Template {
     }
 
     #[inline]
-    pub fn curve_oid(&self) -> untrusted::Input {
+    pub fn curve_oid(&self) -> untrusted::Input<'_> {
         untrusted::Input::from(&self.alg_id_value_()[self.curve_id_index..])
     }
 }
diff --git a/src/polyfill/slice/as_chunks.rs b/src/polyfill/slice/as_chunks.rs
index bfe5ab5..77df415 100644
--- a/src/polyfill/slice/as_chunks.rs
+++ b/src/polyfill/slice/as_chunks.rs
@@ -16,7 +16,7 @@ use super::AsChunksMut;
 use core::ops;
 
 #[inline(always)]
-pub fn as_chunks<T, const N: usize>(slice: &[T]) -> (AsChunks<T, N>, &[T]) {
+pub fn as_chunks<T, const N: usize>(slice: &[T]) -> (AsChunks<'_, T, N>, &[T]) {
     assert!(N != 0, "chunk size must be non-zero");
     let len = slice.len() / N;
     let (multiple_of_n, remainder) = slice.split_at(len * N);
diff --git a/src/polyfill/slice/as_chunks_mut.rs b/src/polyfill/slice/as_chunks_mut.rs
index ae32f04..d24cdbb 100644
--- a/src/polyfill/slice/as_chunks_mut.rs
+++ b/src/polyfill/slice/as_chunks_mut.rs
@@ -15,7 +15,7 @@
 use super::AsChunks;
 
 #[inline(always)]
-pub fn as_chunks_mut<T, const N: usize>(slice: &mut [T]) -> (AsChunksMut<T, N>, &mut [T]) {
+pub fn as_chunks_mut<T, const N: usize>(slice: &mut [T]) -> (AsChunksMut<'_, T, N>, &mut [T]) {
     assert!(N != 0, "chunk size must be non-zero");
     let len = slice.len() / N;
     let (multiple_of_n, remainder) = slice.split_at_mut(len * N);
@@ -52,25 +52,25 @@ impl<T, const N: usize> AsChunksMut<'_, T, N> {
 
     #[cfg(target_arch = "x86_64")]
     #[inline(always)]
-    pub fn as_mut(&mut self) -> AsChunksMut<T, N> {
+    pub fn as_mut(&mut self) -> AsChunksMut<'_, T, N> {
         AsChunksMut(self.0)
     }
 
     #[inline(always)]
-    pub fn as_ref(&self) -> AsChunks<T, N> {
+    pub fn as_ref(&self) -> AsChunks<'_, T, N> {
         AsChunks::<T, N>::from(self)
     }
 
     // Argument moved from runtime argument to `const` argument so that
     // `CHUNK_LEN * N` is checked at compile time for overflow.
     #[inline(always)]
-    pub fn chunks_mut<const CHUNK_LEN: usize>(&mut self) -> AsChunksMutChunksMutIter<T, N> {
+    pub fn chunks_mut<const CHUNK_LEN: usize>(&mut self) -> AsChunksMutChunksMutIter<'_, T, N> {
         AsChunksMutChunksMutIter(self.0.chunks_mut(CHUNK_LEN * N))
     }
 
     #[cfg(target_arch = "x86_64")]
     #[inline(always)]
-    pub fn split_at_mut(&mut self, mid: usize) -> (AsChunksMut<T, N>, AsChunksMut<T, N>) {
+    pub fn split_at_mut(&mut self, mid: usize) -> (AsChunksMut<'_, T, N>, AsChunksMut<'_, T, N>) {
         let (before, after) = self.0.split_at_mut(mid * N);
         (AsChunksMut(before), AsChunksMut(after))
     }
diff --git a/src/rsa/public_modulus.rs b/src/rsa/public_modulus.rs
index c40e49e..77c41a7 100644
--- a/src/rsa/public_modulus.rs
+++ b/src/rsa/public_modulus.rs
@@ -91,3 +91,3 @@ impl PublicModulus {
-    pub(super) fn value(&self, cpu_features: cpu::Features) -> bigint::Modulus<N> {
+    pub(super) fn value(&self, cpu_features: cpu::Features) -> bigint::Modulus<'_, N> {
         self.value.modulus(cpu_features)
     }
-- 
2.34.1
